condensation/cloud.py

```diff
@@ -0,0 +1,366 @@
+        self.improve_accuracy = False
+        LOG.debug("created cloud obj with name " + name)
+
+    def add_nodes(self, nodes_dict):
+        """
+            This method adds nodes to self
+        """
+        LOG.debug("adding nodes to cloud " + self.name)
+        self.nodes = nodes_dict
+        for node in nodes_dict.values():
+            node.cloud = self
+
+    def calc_required_flavors_for_nodes(self):
+        """
+        This method recalculates flavor distribution over nodes for all
+        node names in self.node_ids_to_be_recalculated
+        This trick reduces total complexity of the program
+        We don't have to recalculate distribution for all nodes
+        only for ones that are in array
+        """
+        # we need count for each flavor distribution on nodes that are not full
+        LOG.debug("starting recalculation of flavor distribution over nodes")
+        flavors_dict = {}
+        for node in self.nodes.values():
+            if node.is_full:
+                continue
+            for vm in node.vms.values():
+                if vm.flavor not in flavors_dict:
+                    flavors_dict[vm.flavor] = 0
+                flavors_dict[vm.flavor] += 1
+
+        # just in case - make list of nodes to be recalculated distinct
+        self.node_ids_to_be_recalculated = list(set(self.node_ids_to_be_recalculated))
+        LOG.info("recalculating nodes " + ",".join(self.node_ids_to_be_recalculated))
+        for node_name in self.node_ids_to_be_recalculated:
+            # paranoid check
+            if node_name not in self.nodes:
+                if node_name in self.required_flavors_for_nodes:
+                    del self.required_flavors_for_nodes[node_name]
+                continue
+            node_obj = self.nodes[node_name]
+            if node_obj.vms and not node_obj.is_full:
+                LOG.debug("recalculating " + node_name)
+                self.required_flavors_for_nodes.update({
+                    node_name: node_obj.calculate_flavors_required(
+                        flavors_dict,
+                        self.improve_accuracy)})
+
+        # after updating distribution clear list
+        self.node_ids_to_be_recalculated = []
+        LOG.debug("finished recalculation of flavor distribution over nodes")
+
+    def condense(self, improve_accuracy=False):
+        """
+            This method finds vms distribution on nodes with highest density
+            it runs recursively until we cannot find better solution
+        """
+        self.required_flavors_for_nodes = {}
+        self.improve_accuracy = improve_accuracy
+        self.node_ids_to_be_recalculated = []
+        # recalculate all nodes that are neither full, nor empty
+        for node_name, node_obj in self.nodes.items():
+            if node_obj.vms and not node_obj.is_full:
+                self.node_ids_to_be_recalculated.append(node_name)
+        self.condense_recursively()
+
+    def condense_recursively(self):
```

Comment:
1. Rename all occurrences of "fulled" with "filled" (try google translate to find out what's "fulled" ;))
2. The method looks overly complex. Is there any chance simplifying it?.. Maybe splitting it into multiple self-descriptive method calls would help making it more readable and understandable.